File: JavaFX-Client/src/sample/ConnectionService.java
package sample;

import sample.Model.HeaderType;
import sample.Model.Headers;
import sample.Model.Message;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.SocketChannel;
import java.util.*;

class ConnectionService {

    private SocketChannel socketChannel;


    private boolean connected;
    private int HEADER_LENGTH = 12;
    private int BYTES_TO_READ_LENGTH = 4;
    private String loggedUser;

    ConnectionService(String ip, int port, String name) {
        this.loggedUser = name;
        connected = startConnection(ip, port);
    }

    private boolean startConnection(String ip, int port) {

        boolean success;
        try {
            InetSocketAddress address = new InetSocketAddress(InetAddress.getByName(ip), port);

            socketChannel = SocketChannel.open(address);
            socketChannel.configureBlocking(false);

            success = handShakeServer();
            requestTablesInfo();
        } catch (Exception e) {
            return false;
        }

        return success;
    }

    public void requestTablesInfo() throws IOException {
        send("LIST__TABLES0000");
    }

    private boolean handShakeServer() throws IOException {
        sendWelcomePackage();
        HeaderType header = readHeader();
        read(BYTES_TO_READ_LENGTH);
        return header.equals(HeaderType.WELCOME_USER);
    }


    private String read(int bytesToRead) throws IOException {
        return new String(readByteArray(bytesToRead));
    }

    private byte[] readByteArray(int expectedLength) throws IOException {
        int read = 0;
        int consumed;
        byte[] result = new byte[expectedLength];

        do {
            ByteBuffer buff = ByteBuffer.allocate(expectedLength - read);
            buff.clear();
            consumed = socketChannel.read(buff);
            if(consumed > 0) {
                System.arraycopy(buff.array(), 0, result, read, consumed);
                read += consumed;
            }
            else if (consumed == -1) {
                throw new IOException("Error when reading byte array");
            }

        } while (read != expectedLength);

        return result;
    }

    private void sendWelcomePackage() throws IOException {
        String welcomePackage = "HELLO_SERVER" + String.format("%04d", this.loggedUser.length()) +
                this.loggedUser;
        socketChannel.write(ByteBuffer.wrap(welcomePackage.getBytes()));
    }

    private void send(String message) throws IOException {
        socketChannel.write(ByteBuffer.wrap(message.getBytes()));
    }


    void sendMessage(String message) throws IOException {
        String stringBuilder = "SEND_MESSAGE" + String.format("%04d", message.length()) +
                message;
        send(stringBuilder);
    }

    Message getMessage() throws IOException {
        int length = Integer.parseInt(read(BYTES_TO_READ_LENGTH));
        String userNameAndMessage = read(length);
        return new Message(userNameAndMessage.split("\n")[0], userNameAndMessage.split("\n")[1]);
    }

    boolean isConnected() {
        return connected;
    }

    void closeSocket() throws IOException {
        socketChannel.close();
    }

    HeaderType readHeader() throws IOException {
        String header = read(HEADER_LENGTH);

        if (Headers.HEADERS.containsKey(header)) {
            return Headers.HEADERS.get(header);
        }

        return HeaderType.UNDEFINED;
    }

    private void sendByteArray(byte[] byteArray) throws IOException {
        int length = byteArray.length;
        int sent = 0;
        ByteBuffer buffer = ByteBuffer.wrap(byteArray);

        while(sent != length) {
            sent += socketChannel.write(buffer);

            if(sent != length) {
                byte[] rest = new byte[length - sent];
                buffer.get(rest, sent, length - sent);
                buffer = ByteBuffer.wrap(rest);
            }
        }
    }

    String getGameWaiting() throws IOException {
        read(BYTES_TO_READ_LENGTH);
        return "The game has not started yet";
    }

    String getWinner() throws IOException {
        return "Correct Answer! " + getMessage().getContent();
    }

    String getGameReady() throws IOException {
        int length = Integer.parseInt(read(BYTES_TO_READ_LENGTH));
        String result = read(length);
        return "The drawer is - " + result;
    }

    public BufferedImage getCanvas() throws IOException {
        int length = Integer.parseInt(read(BYTES_TO_READ_LENGTH * 2));
        byte[] bitmapByteArray = readByteArray(length);
        ByteArrayInputStream bais = new ByteArrayInputStream(bitmapByteArray);
        return ImageIO.read(bais);
    }

    public String getThingToDraw() throws IOException {
        int length = Integer.parseInt(read(BYTES_TO_READ_LENGTH));
        String result = read(length);
        return "Your turn! Draw - " + result;
    }

    public List<String> getActiveTables() throws IOException {
        int amountOfTables = Integer.parseInt(read(BYTES_TO_READ_LENGTH));
        String result = read(amountOfTables);
        return Arrays.asList(result.split("\n"));
    }

    public void sendPicture(byte[] picture) throws Exception {
        if(isConnected()){
            String header = "SET___CANVAS" + preparePictureLength(picture.length);
            byte[] byteHeader = concatByteArrays(header.getBytes(), picture);
            sendByteArray(byteHeader);
        }
    }

    public String preparePictureLength(int length) throws Exception {
        if (length > 100000000) throw new Exception("The length of picture is too big!");
        String message = String.format("%08d", length);
        return message;
    }

    private byte[] concatByteArrays(byte[] array1, byte[] array2) {
        byte[] afterConcat = new byte[array1.length + array2.length];
        System.arraycopy(array1, 0, afterConcat, 0, array1.length);
        System.arraycopy(array2, 0, afterConcat, array1.length, array2.length);

        return afterConcat;
    }


    public void createNewTable() throws IOException {
        send("CREATE_TABLE0000");
    }

    public void getTableID() throws IOException {
        int length = Integer.parseInt(read(BYTES_TO_READ_LENGTH));
        int id = Integer.parseInt(read(length));
    }

    public void enterTable(String table) throws IOException {
        String message = "ENTER__TABLE" + String.format("%04d", table.length()) +
                table;
        send(message);
    }

    public void leaveTable() throws IOException {
        send("COMEOUTTABLE0000");
    }

    public void surrender() throws IOException {
        send("GIVE_UP_GAME0000");
    }

    public void handleAbortion() throws IOException {
        int length = Integer.parseInt(read(BYTES_TO_READ_LENGTH));
        String pass = read(length);
    }

    public void handleFail() throws IOException {
        read(BYTES_TO_READ_LENGTH);
    }

    public void pingPong() throws IOException {
        read(BYTES_TO_READ_LENGTH);
        send("PONG____PONG0000");
    }

    public void getStatistics() throws IOException {
        send("GETSTATISTIC0000");
    }


    public Map<String,String> seeStatistics() throws IOException {
        int amountOfStats = Integer.parseInt(read(BYTES_TO_READ_LENGTH));
        String result = read(amountOfStats);
        List<String> array = Arrays.asList(result.split("\n"));
        HashMap<String,String> map = new HashMap<>();
        for(int i = 0; i<array.size(); ++i){
            map.put(array.get(i),array.get(++i));
        }
        return map;
    }
}

File: android_client/app/src/main/java/com/acharchu/charades/ConnectionService.kt
package com.acharchu.charades

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import java.lang.StringBuilder
import java.net.InetAddress
import java.net.InetSocketAddress
import java.nio.ByteBuffer
import java.nio.channels.SocketChannel

class ConnectionService {

    companion object {
        private const val serverHost = "10.0.2.2" //has to be changed if u want to run on real smartphone
        private const val serverPort = 44444
        private var socketChannel: SocketChannel? = null

        private const val HEADER_LENGTH = 12
        private const val BYTES_TO_READ_LENGTH = 4

        public var INTERRUPT = false

        var status = State.DISCONNECTED

        private lateinit var id: String

        fun connectToServer() : Boolean {
            try {
                val address = InetSocketAddress(InetAddress.getByName(serverHost), serverPort)

                socketChannel = SocketChannel.open(address)
                socketChannel!!.configureBlocking(false)

            } catch (exception: Exception) {
                return false
            }

            return true
        }

        private fun sendString(msg: String) {
            socketChannel!!.write(ByteBuffer.wrap(msg.toByteArray()))
        }

        private fun sendByteArray(byteArray: ByteArray) {
            send(byteArray.size, ByteBuffer.wrap(byteArray))
        }

        private fun send(bytesToSend: Int, message: ByteBuffer) {
            var bytesAlreadySent = 0
            var byteBuffer = message

            while(bytesAlreadySent != bytesToSend) {
                bytesAlreadySent += socketChannel!!.write(byteBuffer)

                if(bytesAlreadySent != bytesToSend) {
                    val restOfMessage = ByteArray(bytesToSend - bytesAlreadySent)
                    byteBuffer.get(restOfMessage,bytesAlreadySent,bytesToSend - bytesAlreadySent)
                    byteBuffer = ByteBuffer.wrap(restOfMessage)
                }
            }
        }

        fun setId(name: String) {
            id = name
        }

        fun performServerHandshake() : Boolean {
            sendString("HELLO_SERVER${prepareMessageLength(id.length)}$id")
            val handshakeResult = read(16) //check if result is ok

            if (handshakeResult == "WELCOME_USER0000") {
                status = State.CONNECTED
                return true
            }
            else if (handshakeResult == "COMMANDFAILD0000")
                return false
            else
                return false
        }

        fun joinToTable(table_number: Int) {
            sendString("ENTER__TABLE${prepareMessageLength(table_number.toString().length)}$table_number")
        }

        fun comeOutFromTable() {
            sendString("COMEOUTTABLE0000")
        }

        fun giveUpAGame() {
            sendString("GIVE_UP_GAME0000")
        }

        fun ping_ping() {
            read(BYTES_TO_READ_LENGTH)
        }

        fun pong_pong() {
            sendString("PONG____PONG0000")
        }

        fun getStatistics() {
            sendString("GETSTATISTIC0000")
        }

        fun getBestUsersAndScores() : List<String>? {
            val length = read(BYTES_TO_READ_LENGTH)

            if(length.toInt() != 0) {
                val tableStringList = read(length.toInt()).split('\n')
                return prepareScores(tableStringList.subList(0, tableStringList.size-1))
            } else
                return null

        }

        private fun prepareScores(splittedList: List<String>) : List<String> {
            val list = mutableListOf<String>()

            var i = 0
            while (i != splittedList.size) {
                list.add(splittedList[i] + " : " + splittedList[i+1])
                i += 2
            }

            return list
        }

        fun getIdOfCreatedTable() : Int {
            val length = read(BYTES_TO_READ_LENGTH)
            val id = read(length.toInt())
            return id.toInt()
        }

        fun gameAborted() : String {
            val length = read(BYTES_TO_READ_LENGTH)
            val answer = read(length.toInt())
            return "Game aborted! The right answer - $answer"
        }

        fun commandFailed() {
            read(BYTES_TO_READ_LENGTH)
        }

        fun createTable() {
            sendString("CREATE_TABLE0000")
        }

        fun listAvailableTables() {
            sendString("LIST__TABLES0000")
        }

        fun getTableList(): List<Int>? {
            val length = read(BYTES_TO_READ_LENGTH)

            return if(length.toInt() != 0) {
                val tableStringList = read(length.toInt()).split('\n')
                tableStringList.subList(0, tableStringList.size-1).map { it.toInt() }
            } else
                null
        }


        fun sendMessage(msg: String) {
            sendString("SEND_MESSAGE${prepareMessageLength(msg.length)}$msg")
        }

        private fun prepareMessageLength(length: Int): String {
            val builder = StringBuilder()

            builder.append(length)

            return when {
                length < 10 -> "000$length"
                length < 100 -> "00$length"
                length < 1000 -> "0$length"
                length < 10000 -> length.toString()
                //out of bound error
                else -> ""
            }
        }

        fun getMessage(): String? {

            val usernameAndMessageLength = read(BYTES_TO_READ_LENGTH)
            val usernameAndMessage = read(usernameAndMessageLength.toInt()).split('\n')

            return "${usernameAndMessage[0]}: ${usernameAndMessage[1]}"
        }

        fun getHeader(): HeaderType? {
            val header = read(HEADER_LENGTH)
            print(header)
            if (Headers.headers.containsKey(header))
                return Headers.headers[header]

            return HeaderType.UNDEFINED
        }

        private fun read(bytesToRead: Int): String {
            return String(readByteArray(bytesToRead))
        }

        private fun readByteArray(bytesToRead: Int): ByteArray {
            var amountOfCharacters = 0
            var consumedCharacters = 0
            val result = ByteArray(bytesToRead)

            while (amountOfCharacters != bytesToRead && consumedCharacters != -1 && !INTERRUPT) {  // read z offsetem
                val buffer = ByteBuffer.allocate(bytesToRead - amountOfCharacters)
                consumedCharacters = socketChannel?.read(buffer)!!

                if (consumedCharacters != 0 && consumedCharacters != -1) {
                    System.arraycopy(buffer.array(), 0, result, amountOfCharacters, consumedCharacters)
                    amountOfCharacters += consumedCharacters
                } else if (consumedCharacters == -1)
                    throw Throwable("CONNECTION CLOSED")
            }
            INTERRUPT = false

            return result
        }

        fun skipLeftovers() {
            do {
                val buffer = ByteBuffer.allocate(100)
                val consumedCharacters = socketChannel?.read(buffer)
            } while (consumedCharacters != 0)
        }

        fun closeSocket() {
            socketChannel!!.close()
            status = State.DISCONNECTED
        }

        fun getCanvas() : Bitmap {
            val length = read(BYTES_TO_READ_LENGTH*2)
            val bitmapByteArray = readByteArray(length.toInt())
            return BitmapFactory.decodeByteArray(bitmapByteArray, 0, bitmapByteArray.size)
        }

        fun getGameWaiting(): String {
            readByteArray(BYTES_TO_READ_LENGTH)
            return "The game has not started yet."
        }

        fun getTheWinner(): String {
            return "CORRECT ANSWER! ${getMessage()}"

        }

        fun getGameReady(): String{
            val length = read(BYTES_TO_READ_LENGTH)
            val result = read(length.toInt())

            return "The drawer is - $result"
        }

        fun getThingToDraw() : String {
            val length = read(BYTES_TO_READ_LENGTH)
            val result = read(length.toInt())
            return "Your turn! Draw - $result"
        }

        fun sendPicture(picture: ByteArray) {
            var header = "SET___CANVAS${preparePictureLength(picture.size)}".toByteArray()
            header += picture
            sendByteArray(header)
        }

        private fun preparePictureLength(length: Int): String {
            val builder = StringBuilder()

            builder.append(length)

            return when {
                length < 10 -> "0000000$length"
                length < 100 -> "000000$length"
                length < 1000 -> "00000$length"
                length < 10000 -> "0000$length"
                length < 100000 -> "000$length"
                length < 1000000 -> "00$length"
                length < 10000000 -> "0$length"
                length < 100000000 -> length.toString()
                //out of bound error
                else -> throw Exception("The length of the picture is too big!")
            }

        }
    }
}